File: searching/binarySearch.js
// works only on sorted arrays
// time complexity: O(nlog2(n))
const binarySearch = (array, target) => {
  let low = 0,
    high = array.length - 1;
  while (low <= high) {
    const mid = Math.floor((low + high) / 2);
    if (array[mid] === target) return mid;
    else if (array[mid] > target) high = mid - 1;
    else low = mid + 1;
  }
  return -1;
};

File: searching/binarySearch.py
from typing import List

def binarySearch(nums: List[int], target: int) -> int:
  n = len(nums) 
  low, high = 0, n - 1 
  while low <= high:
    mid = (low + high) // 2
    if nums[mid] == target: return mid
    elif nums[mid] > target: high = low - 1
    else: low = mid + 1
  return -1
# Time complexity: O(log2(n))